### PR TITLE
Implement Ice Bomb

### PR DESCRIPTION
**Pull request by zSALLAZAR**
## Introduction
This is a Education Edition feature.

The ice bomb is an item that is used to freeze water into ice.

## Tests

https://user-images.githubusercontent.com/59490940/207361701-bd0be042-4580-497d-915c-c1ca03e8fecb.mp4

### CODE REVIEW

File: src/entity/EntityFactory.php [hunk 1]
```diff
@@ -112,6 +113,10 @@ public function __construct(){
 			return new FallingBlock(Helper::parseLocation($nbt, $world), FallingBlock::parseBlockNBT(BlockFactory::getInstance(), $nbt), $nbt);
 		}, ['FallingSand', 'minecraft:falling_block'], LegacyIds::FALLING_BLOCK);
 
+		$this->register(IceBomb::class, function(World $world, CompoundTag $nbt) : IceBomb{
+			return new IceBomb(Helper::parseLocation($nbt, $world), null, $nbt);
+		}, ['ThrownIceBomb', 'minecraft:ice_bomb'], LegacyIds::ICE_BOMB);
```
Comment (dktapps) [inline src/entity/EntityFactory.php]: Don't assume these. They need to match the vanilla IDs.

Comment (zSALLAZAR) [reply src/entity/EntityFactory.php]: What else are the correct vanilla ids?

Comment (dktapps) [reply src/entity/EntityFactory.php]: I don't know. That's for you to research.

Comment (IvanCraft623) [reply src/entity/EntityFactory.php]: Vanilla NBT tree:
image
`identifier`: `minecraft:ice_bomb`

Comment (dktapps) [reply src/entity/EntityFactory.php]: There's no evidence for any such thing as `ThrownIceBomb`. Those style of ID used to be used in Java, but Java doesn't have ice bombs.

File: src/entity/projectile/IceBomb.php [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+class IceBomb extends Throwable{
+	public static function getNetworkTypeId() : string{ return EntityIds::ICE_BOMB; }
+
+	public function getResultDamage() : int{
+		return -1;
+	}
+
+	protected function calculateInterceptWithBlock(Block $block, Vector3 $start, Vector3 $end) : ?RayTraceResult{
+		if($block instanceof Water){
```
Comment (ipad54) [inline src/entity/projectile/IceBomb.php]: I think you can use getTypeId() for that

Comment (zSALLAZAR) [reply src/entity/projectile/IceBomb.php]: Why should I?

Comment (dktapps) [reply src/entity/projectile/IceBomb.php]: `Water` class could be reused for other stuff which isn't supposed to behave the same way, but in general, `instanceof` should be avoided. It's not a safe way to determine a block's type.

File: src/entity/projectile/IceBomb.php [hunk 3]
```diff
@@ -44,7 +43,7 @@ public function getResultDamage() : int{
 	}
 
 	protected function calculateInterceptWithBlock(Block $block, Vector3 $start, Vector3 $end) : ?RayTraceResult{
-		if($block instanceof Water){
+		if($block->getTypeId() === VanillaBlocks::WATER()->getTypeId()){
```
Comment (dktapps) [inline src/entity/projectile/IceBomb.php]:
```suggestion
		if($block->getTypeId() === BlockTypeIds::WATER){
```

File: src/entity/projectile/IceBomb.php [hunk 4]
```diff
@@ -80,7 +79,7 @@ protected function onHitBlock(Block $blockHit, RayTraceResult $hitResult) : void
 		for($x = $posX - 1; $x <= $posX + 1; $x++){
 			for($y = $posY - 1; $y <= $posY + 1; $y++){
 				for($z = $posZ - 1; $z <= $posZ + 1; $z++){
-					if($world->getBlockAt($x, $y, $z) instanceof Water){
+					if($world->getBlockAt($x, $y, $z)->getTypeId() === VanillaBlocks::WATER()->getTypeId()){
```
Comment (dktapps) [inline src/entity/projectile/IceBomb.php]:
```suggestion
					if($world->getBlockAt($x, $y, $z)->getTypeId() === BlockTypeIds::WATER){
```

**Comment by dktapps** [comment]
There are some merge errors that need fixing

File: src/item/StringToItemParser.php [hunk 5]
```diff
@@ -1299,6 +1299,7 @@ private static function registerItems(self $result) : void{
 		$result->register("heart_of_the_sea", fn() => Items::HEART_OF_THE_SEA());
 		$result->register("honey_bottle", fn() => Items::HONEY_BOTTLE());
 		$result->register("honeycomb", fn() => Items::HONEYCOMB());
+		$result->register('ice_bomb', fn() => Items::ICE_BOMB());
```
Comment (dktapps) [inline src/item/StringToItemParser.php]: This should use double quotes

File: src/item/VanillaItems.php [hunk 6]
```diff
@@ -447,6 +448,7 @@ protected static function setup() : void{
 		self::register("heart_of_the_sea", new Item(new IID(Ids::HEART_OF_THE_SEA), "Heart of the Sea"));
 		self::register("honey_bottle", new HoneyBottle(new IID(Ids::HONEY_BOTTLE), "Honey Bottle"));
 		self::register("honeycomb", new Item(new IID(Ids::HONEYCOMB), "Honeycomb"));
+		self::register('ice_bomb', new IceBomb(new IID(Ids::ICE_BOMB), 'Ice Bomb'));
```
Comment (dktapps) [inline src/item/VanillaItems.php]: Same again here

**Review by jasonw4331 — APPROVED** [review]

**Review by ShockedPlot7560 — DISMISSED** [review]

**Comment by pandaaaBE** [comment]
this pr should target minor-next

a little change would be required, make getNetworkTypeId() static

**Comment by zSALLAZAR** [comment]
Changing the target branch to minor-next adds commits from major-next to this PR and Idk how to remove them

**Comment by dktapps** [comment]
I think this is done, I made a few minor alterations and rebased it, but code was mostly OK anyway

**Pull request merged**